autopaho/backoff.go

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright (c) 2024 Contributors to the Eclipse Foundation
+ *
+ *  All rights reserved. This program and the accompanying materials
+ *  are made available under the terms of the Eclipse Public License v2.0
+ *  and Eclipse Distribution License v1.0 which accompany this distribution.
+ *
+ * The Eclipse Public License is available at
+ *    https://www.eclipse.org/legal/epl-2.0/
+ *  and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ *  SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+
+package autopaho
+
+import (
+	"math/rand"
+	"time"
+)
+
+// The BackoffStrategy is a container for the configuration of a given strategy whereas the returned Backoff is the actual implementation
+// As such BackoffStrategy instances are supposed to be thread safe and freely sharable between users.
+type BackoffStrategy interface {
```

Comment:
I'm suggesting fairly major changes to handle an issue you were not aware of so feel free to tell me to get lost!

Consider the following possibility:
1. Establish connection and send CONNECT (so `establishServerConnection` has completed)
2. Broker rejects CONNECT (say bad password) so look to 1

This can result in us hammering the broker with CONNECT packets (not a good look!). It would be nice if we could solve this issue at the same time as dealing with the backoff on the network level connection.

I've had a quick think about options and wonder if the following might work (and provide a solution that's a bit simpler than your approach):

```
ConnectionBackoff func(ctx Context.Context, establishConnection func(ctx Context.Context) error) error // Passed function that will establish network connection and calls it until a `nil` error is returned or the context is cancelled.
```

Then in `establishServerConnection` you would do something like:

```
var cli *paho.Client,
var connack *paho.Connack)

err := cfg.ConnectionBackoff( ctx, func(ctx Context.Context) error {
     for _, u := range cfg.ServerUrls {
		connectionCtx, cancelConnCtx := context.WithTimeout(ctx, cfg.ConnectTimeout)
....
    }
}
if err != nil { // this would only happen when context cancelled

}
return cli, connack
```

The advantage of this is that it means that the `ConnectionBackoff ` has complete control of the process (including resetting, or not, between connections). `ConstantBackoff` could be very simple (pseudo code!):

```
type ConstantBackoff struct {
	delay       time.Duration
	lastAttempt time.Time
}

func (cb *ConstantBackoff) Run(ctx context.Context, establishConnection func(ctx Context.Context) error) {
	// The previous connection attempt may have been very recent; we want at least delay between attempts
	select {
	case time.After(cb.delay - time.Since(cb.lastAttempt)): // Might be negative; i.e. no delay
	case <-ctx.Done():
		return ctx.Err()
	}

	for {
		err := establishConnection(ctx)
		if err == nil || ctx.Err() != nil {
			return err
		}
		select {
		case <-time.After(cb.delay):
		case <-ctx.Done():
			return ctx.Err()
		}
	}
}
```

Other backoffs could be a bit more complex (and take into account whatever factors are relevant).

Reply:
My approach was to have a simple implementation. Therefore one can re-use / share the `BackoffStrategy` while the actual `Backoff` is "throwaway". It doesn't need a "reset" of anything fancy.
In my experience these mechanisms, where the server could potentially say: "come back in one hour", sound nice but are not worth the hassle. Often the "one hour" is either to early or to late.
Having a rate limiter on the server side and an exponential backoff on the client side, has worked for me with planned and unplanned outages in the past. But that's only my use case... so your mileage might vary.
One aspect which 
Either way, IMO something like taking care of CONNECT should have a dedicated PR and discussion (as it is seems to me way more complex than the backoff itself).

Reply:
> IMO something like taking care of CONNECT should have a dedicated PR and discussion (as it seems to me way more complex than the backoff itself).

Sorry - I made an error here (the issue with working on multiple clients concurrently!). In this client `establishServerConnection` only returns after the `CONNACK` is received. As such this is not the issue I thought it was (and your current strategy would work).

Note that the reason I raised this (and want to put some time into considering this change) is due to experience with the V3 client where we hit issues because the options became very complicated, so I'm careful with new options in this project (once an option is added it's very hard to remove!).

>where the server could potentially say: "come back in one hour",

The one thing that may be a little bit like the scenario you mention ("come back later") is duplicate client id connect loops. This is very common for users new to the protocol where they have two clients with the same client ID and end up with a connect loop (two connections "fighting"). With v3.1 this was difficult to detect (the server just dropped the connection) but with V5 a specific Disconnect would be sent (Reason Code 0x8E - Session taken over) so there is a potential that we could handle it in the future. I only mention this as being able to prevent connection storms when this is happening would be beneficial longer term (so would be nice to have the option of adding support for this).

>My approach was to have a simple implementation. Therefore one can re-use / share the BackoffStrategy while the actual Backoff is "throwaway". It doesn't need a "reset" of anything fancy.

I understand that; however part of my reaction here might have been because your code does look a bit Java'ish. I avoided this in the last set of comments because it's highly subjective; but as you specifically asked "So if I hit any no-Go-s" here are a few thoughts (I'll add a few more comments in the code):

* The structure made me think of the Java factory pattern; obviously this pattern is used in Go but it's a lot less common than in Java.
* Ive not seen use of underscore prefixes (i.e.  `_Duration`) other than in unexported globals. I'd expect this variable to be called `delay` or similar (with the lower case `d` indicating it's private).

My suggestion would be to simplify this to (using the `Reset` option you felt complicated things):

```
// backoff represents a configured instance of the strategy which computes the next backoff duration.
type backoff interface {
        Reset() 
	NextBackOff() time.Duration
}
```

There are a few reasons for this:

It's simple.

It's compatible with [`cenkalti/backoff`](https://github.com/cenkalti/backoff) which is the most well known back-off library I'm aware of (so would provide a range of options users could simply plug in). Note: I would prefer to avoid a dependency so the basic strategies would be implemented in this library (which you have done).

It allows authors to consider previous state (e.g. never attempt to connect twice in two minutes) which is doable, but more difficult, with the current structure.

We can extend it in the future without breaking users code. For example to handle the connection fight situation mentioned above we could add:

``` 
type backoffWithDisconnected interface {
     backoff 
    Disconnected(*package.Disconnect) // called when Disconnect packet received
}
```

Then, when a disconnect packet is received, we would do something like:

```
if on, dwd := cfg.Backoff.(backoffWithDisconnected) ; ok {
   dwd.Disconnected(pkt)
}
```

This is just an off the cuff thought; so treat as pseudo code but aims to show why this provided more flexibility.

Note: Once more I'm not saying that this is the right approach but do want to be careful with this because once an interface is selected I think we need to consider it locked in.

Reply:
> I understand that; however part of my reaction here might have been because your code does look a bit Java'ish.

You are totally right. Funny how it shows. I've been doing Java for many years and as you correctly guessed this code originates from Java. In Java it was very simple:

```
public interface BackoffStrategy {
    Iterator<Duration> backoff();
}
```

As the custom `Iterator` was a nested class in the implementation it could easily/safely resuse final fields from the outer class.

But somehow I was not able to implement something similar simple in Go.